File: demo/runAltNewtonCD.m
function [Lambda, Theta, stats] = runAltNewtonCD(...
    Y, X, lambdaLambda, lambdaTheta, options)
% Inputs:
% Y (n samples x q dimensions)
% X (n samples x p dimensions)
% lambdaLambda (regularization for Lambda)
% lambdaTheta (regularization for Theta)
% [options] struct with the following options:
%   - verbose(0): show information or not (0 or 1)
%   - max_outer_iters(50): max number of outer iterations
%   - sigma(1e-4): backtracking termination criterion
%   - tol(1e-2): tolerance for terminating outer loop
%   - Lambda0(none): q x q sparse matrix to initialize Lambda
%   - Theta0(none): p x q sparse matrix to initialize Theta
%   - refit(0): refit selected model without adding any edges
    
    olddir = pwd;
    thisfunc = which(mfilename());
    thisdir = thisfunc(1:end-16);
    cd(thisdir);

    L0_str = '';
    T0_str = '';
    dummy = randi(1e6);
    verbose = 0;
    max_outer_iters = 50;
    sigma = 1.0e-4;
    tol = 1.0e-2;
    refit = 0;

    if exist('options', 'var')
        if isfield(options, 'verbose')
            verbose = options.verbose;
        end
        if isfield(options, 'max_outer_iters')
            max_outer_iters = options.max_outer_iters;
        end
        if isfield(options, 'tol')
            tol = options.tol;
        end
        if isfield(options, 'refit')
            refit = options.refit;
        end
        if isfield(options, 'Lambda0')
            Lambda0file = sprintf('Lambda0-dummy-%i.txt', dummy);
            sparse_to_txt(Lambda0file, options.Lambda0);
            L0_str = sprintf('-L \"%s\" ', Lambda0file);
        end
        if isfield(options, 'Theta0')
            Theta0file = sprintf('Theta0-dummy-%i.txt', dummy);
            sparse_to_txt(Theta0file, options.Theta0);
            T0_str = sprintf('-T \"%s\" ', Theta0file);
        end
    end
    [n_y, q] = size(Y);
    [n_x, p] = size(X);

    Yfile = sprintf('Y-dummy-%i.txt', dummy);
    Xfile = sprintf('X-dummy-%i.txt', dummy);
    Lambdafile = sprintf('Lambda-dummy-%i.txt', dummy);
    Thetafile = sprintf('Theta-dummy-%i.txt', dummy);
    statsfile = sprintf('stats-dummy-%i.txt', dummy);
    dlmwrite(Yfile, Y, 'delimiter', ' ', 'precision', 10);
    dlmwrite(Xfile, X, 'delimiter', ' ', 'precision', 10);
    option_str = sprintf('-y %f -x %f -v %i -i %i -s %f -q %f -r %i', ...
        lambdaLambda, lambdaTheta, ...
        verbose, max_outer_iters, sigma, tol, refit);
    command_str = sprintf('%s %s %s %s %i %i %i %i %s %s %s %s %s', ...
        '../AltNewtonCD/cggmfast_run', option_str, ...
        L0_str, T0_str, ...
        n_x, p, n_y, q, Yfile, Xfile, ...
        Lambdafile, Thetafile, statsfile);

    system(command_str);
    Lambda = txt_to_sparse(Lambdafile);
    Theta = txt_to_sparse(Thetafile);
    stats = txt_to_struct(statsfile);
    system(['rm ' Yfile ' ' Xfile ' ' Lambdafile ' ' Thetafile ' ' statsfile]);
    if L0_str
        system(['rm ' Lambda0file]);
    end
    if T0_str
        system(['rm ' Theta0file]);
    end
    cd(olddir);
end

File: demo/runAltNewtonCD.py
#!/bin/python

import csv
import os
import subprocess
import sys
import matplotlib.pyplot as plt
import numpy as np
import numpy.random as rnd
import scipy.sparse as ssp

def read_stats(f):
    stats_dict = {}
    with open(f, "r") as sf:
        sfr = csv.reader(sf, delimiter=" ")
        for row in sfr:
            stats_dict[row[0]] = np.array([float(x) for x in row[1:] if x])
    return stats_dict

def txt_to_sparse(filename):
    with open(filename, "r") as tf: 
        tfr = csv.reader(tf, delimiter=" ")
        first_row = tfr.next()
        p = int(first_row[0])
        q = int(first_row[1])
        nnz = int(first_row[2])
        the_rest = [(int(row[0])-1, int(row[1])-1, float(row[2]))
                    for row in tfr]
    (i, j, data) = zip(*the_rest)
    Theta = ssp.coo_matrix((list(data), (list(i),list(j))), shape=(p, q))
    return Theta

def runAltNewtonCD(
        Y, X, lambdaLambda, lambdaTheta, 
        verbose=False, max_outer_iters=50, sigma=1e-4, tol=1e-2, refit=False):
    '''
    Inputs:
    Y (n samples x q dimensions)
    X (n samples x p dimensions)
    lambdaLambda (regularization for Lambda)
    lambdaTheta (regularization for Theta)
    verbose: print information or not
    max_outer_iters: max number of outer iterations
    sigma: backtracking termination criterion
    tol: tolerance for terminating outer loop
    refit: refit selected model without adding any edges
    '''

    olddir = os.getcwd()
    thisdir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(thisdir) # move to demo directory

    dummy = rnd.randint(low=0, high=1e6)

    (n_y, q) = Y.shape
    (n_x, p) = X.shape

    Yfile = "Y-dummy-%i.txt" % (dummy)
    Xfile = "X-dummy-%i.txt" % (dummy)
    Lambdafile = "Lambda-dummy-%i.txt" % (dummy)
    Thetafile = "Theta-dummy-%i.txt" % (dummy)
    statsfile = "stats-dummy-%i.txt" % (dummy)
    np.savetxt(Yfile, Y, fmt="%.10f", delimiter=" ")
    np.savetxt(Xfile, X, fmt="%.10f", delimiter=" ")
    option_str = "-y %f -x %f -v %i -i %i -s %f -q %f -r %i" % (
        lambdaLambda, lambdaTheta,
        verbose, max_outer_iters, sigma, tol, refit)
    command_str = "%s %s %i %i %i %i %s %s %s %s %s" % (
        "../AltNewtonCD/cggmfast_run", option_str,
        n_x, p, n_y, q, Yfile, Xfile,
        Lambdafile, Thetafile, statsfile)

    ret = os.system(command_str)
    Lambda = txt_to_sparse(Lambdafile)
    Theta = txt_to_sparse(Thetafile)
    stats = read_stats(statsfile)
    rmline = "rm %s %s %s %s %s" % (Yfile, Xfile, Lambdafile, Thetafile, statsfile)
    ret = os.system(rmline)
    os.chdir(olddir)
    return (Lambda, Theta)

if __name__ == "__main__":
    thisdir = os.path.dirname(os.path.abspath(__file__))
    print thisdir
    X = np.loadtxt(thisdir + "/Xfile")
    Y = np.loadtxt(thisdir + "/Yfile")
    lambdaLambda = 0.1
    lambdaTheta = 0.2
    (Lambda, Theta) = runAltNewtonCD(Y, X, lambdaLambda, lambdaTheta)
    plt.figure()
    plt.spy(Theta)
    plt.title("Theta")
    plt.figure()
    plt.spy(Lambda)
    plt.title("Lambda")
    plt.figure()
    plt.imshow(Theta.todense(), interpolation="nearest")
    plt.title("Theta dense")
    plt.colorbar()
    plt.show()
